File: polyglot/csharp/manifest_parser.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace baadiff
{
    /// <summary>
    /// HIPAA Security Rule Control mappings for Kubernetes manifests.
    */
    internal static class HipaaControls
    {
        public const string ACCESS_CONTROL = "AC";
        public const string AUDIT_CONTROLS = "AU";
        public const string INTEGRITY_CONTROLS = "IA";
        public const string PHYSICAL_CONTROLS = "PE";
        public const string NETWORK_CONTROLS = "NE";
        public const string POLICY_MANAGEMENT = "PO";

        // Sample control IDs (Simplified for demo)
        private static readonly Dictionary<string, string> ControlMap = new()
        {
            ["rbac:admin"] = $"{ACCESS_CONTROL}:AC-2",      // Unique user IDs
            ["rbac:service-account"] = $"{ACCESS_CONTROL}:AC-3",  // MFA/least privilege
            ["network:ingress-tls"] = $"{NETWORK_CONTROLS}:NE-4",   // Encryption in transit
            ["logging:audit-logs"] = $"{AUDIT_CONTROLS}:AU-2",      // Logging
            ["secrets:vault-ref"] = $"{INTEGRITY_CONTROLS}:IA-5",   // Integrity validation
        };

        public static string GetControlId(string key) => ControlMap.TryGetValue(key, out var id) ? id : null;
    }

    /// <summary>
    /// Represents a single HIPAA control finding.
    */
    internal class Finding
    {
        public string ControlId { get; set; } = "";
        public string Category => HipaaControls.GetControlId(ControlId) ?? "OTHER";
        public string ResourcePath { get; set; } = "";
        public string Description { get; set; } = "";
        public bool Passes { get; set; }
        public List<string> Evidence { get; set; } = new();

        public override string ToString() => $"{ControlId}: {(Passes ? "PASS" : "FAIL")} - {Description}";
    }

    /// <summary>
    /// Represents a parsed Kubernetes manifest resource.
    */
    internal class K8sResource
    {
        public string Kind { get; set; } = "";
        public string Name { get; set; } = "";
        public Dictionary<string, object>? Metadata { get; set; }
        public Dictionary<string, object>? Spec { get; set; }

        public static K8sResource? Parse(string yamlContent)
        {
            // Simple YAML parser for demo - in production use YamlDotNet or System.Text.Json with custom converters
            var json = JsonSerializer.Deserialize<JsonElement>(yamlContent);
            
            if (json.ValueKind != JsonValueKind.Object || 
                !json.TryGetProperty("kind", out var kindProp) || 
                !json.TryGetProperty("metadata", out var metaProp))
            {
                return null;
            }

            return new K8sResource
            {
                Kind = kindProp.GetString() ?? "",
                Name = metaProp["name"]?.GetString() ?? "",
                Metadata = metaProp.Deserialize<Dictionary<string, object>>()
            };
        }
    }

    /// <summary>
    /// Main HIPAA manifest scanner.
    */
    internal class HipaaScanner
    {
        private readonly List<K8sResource> _resources = new();
        private readonly List<Finding> _findings = new();

        public void AddManifest(string yamlContent)
        {
            var resource = K8sResource.Parse(yamlContent);
            if (resource != null)
                _resources.Add(resource);
        }

        public void Scan()
        {
            // 1. Access Control Checks
            CheckAccessControl();
            
            // 2. Audit Controls Checks  
            CheckAuditControls();
            
            // 3. Network/Encryption Checks
            CheckNetworkSecurity();
            
            // 4. Secret Management
            CheckSecretManagement();

            // Remove duplicates and sort findings
            _findings = _findings.DistinctBy(f => f.ControlId).ToList();
        }

        private void CheckAccessControl()
        {
            var adminRoles = _resources.Where(r => 
                r.Spec?.TryGetValue("rules", out var rules) == true &&
                rules.TryGetProperty("apiGroups", out var apiGroups) &&
                apiGroups.GetString() == "rbac.authorization.k8s.io" &&
                rules.TryGetProperty("verbs", out var verbs) &&
                verbs.GetString() == "admin")
                   .Select(r => r.Name);

            if (adminRoles.Any())
            {
                _findings.Add(new Finding
                {
                    ControlId = $"{HipaaControls.ACCESS_CONTROL}:AC-2",
                    ResourcePath = string.Join(", ", adminRoles),
                    Description = "Potential over-permissive RBAC admin roles detected. Review for least privilege.",
                    Passes = false,
                    Evidence = new() { "Found in: " + string.Join(", ", adminRoles) }
                });
            }

            var serviceAccounts = _resources.Where(r => r.Kind == "ServiceAccount");
            if (serviceAccounts.Any(sa => !sa.Spec?.TryGetValue("secrets", out _) ?? false))
            {
                _findings.Add(new Finding
                {
                    ControlId = $"{HipaaControls.ACCESS_CONTROL}:AC-3",
                    ResourcePath = string.Join(", ", serviceAccounts.Select(r => r.Name)),
                    Description = "Service accounts without secrets. Consider adding token request for MFA-like auth.",
                    Passes = false,
                    Evidence = new() { "Service accounts: " + string.Join(", ", serviceAccounts.Select(r => r.Name)) }
                });
            }
        }

        private void CheckAuditControls()
        {
            var loggingEnabled = _resources.Any(r => 
                r.Kind == "Pod" && 
                r.Spec?.TryGetValue("containers", out var containers) == true &&
                containers.TryGetProperty("image", out _) &&
                (string)r.Spec["containers"]["image"]!.Contains("fluentd") ||
                (string)r.Spec["containers"]["image"]!.Contains("logstash"));

            if (!loggingEnabled)
            {
                _findings.Add(new Finding
                {
                    ControlId = $"{HipaaControls.AUDIT_CONTROLS}:AU-2",
                    ResourcePath = "Cluster-wide",
                    Description = "No fluentd/logstash detected. Audit logging may be incomplete.",
                    Passes = false,
                    Evidence = new() { "Check cluster logging configuration" }
                });
            }

            var auditAnnotations = _resources.Where(r => 
                r.Metadata?.TryGetValue("annotations", out var ann) == true &&
                ann.TryGetProperty("audit-log-enabled", out _) ||
                ann.TryGetProperty("hipaa-audit", out _));

            if (auditAnnotations.Any())
            {
                _findings.Add(new Finding
                {
                    ControlId = $"{HipaaControls.AUDIT_CONTROLS}:AU-3",
                    ResourcePath = string.Join(", ", auditAnnotations.Select(r => r.Name)),
                    Description = "Audit annotations detected - verify they're being consumed.",
                    Passes = true,
                    Evidence = new() { "Found in: " + string.Join(", ", auditAnnotations.Select(r => r.Name)) }
                });
            }
        }

        private void CheckNetworkSecurity()
        {
            var tlsEnabled = _resources.Any(r => 
                r.Kind == "Ingress" &&
                r.Spec?.TryGetValue("tls", out _) == true);

            if (!tlsEnabled)
            {
                _findings.Add(new Finding
                {
                    ControlId = $"{HipaaControls.NETWORK_CONTROLS}:NE-4",
                    ResourcePath = "Cluster-wide Ingress",
                    Description = "Ingress resources may not have TLS configured. Check for plain HTTP.",
                    Passes = false,
                    Evidence = new() { "Review all Ingress CRDs" }
                });
            }

            var networkPolicies = _resources.Where(r => r.Kind == "NetworkPolicy");
            if (networkPolicies.Any())
            {
                _findings.Add(new Finding
                {
                    ControlId = $"{HipaaControls.NETWORK_CONTROLS}:NE-6",
                    ResourcePath = string.Join(", ", networkPolicies.Select(n => n.Name)),
                    Description = "Network policies found. Verify they implement segmentation.",
                    Passes = true,
                    Evidence = new() { "Found in: " + string.Join(", ", networkPolicies.Select(n => n.Name)) }
                });
            }
        }

        private void CheckSecretManagement()
        {
            var vaultRefs = _resources.Where(r => 
                r.Spec?.TryGetValue("secrets", out _) == true ||
                r.Metadata?.TryGetValue("annotations", out _) == true &&
                (string?)r.Metadata["annotations"]?["vault.hashicorp.com/agent-inject"] != null);

            if (vaultRefs.Any())
            {
                _findings.Add(new Finding
                {
                    ControlId = $"{HipaaControls.INTEGRITY_CONTROLS}:IA-5",
                    ResourcePath = string.Join(", ", vaultRefs.Select(r => r.Name)),
                    Description = "Secrets managed via HashiCorp Vault detected.",
                    Passes = true,
                    Evidence = new() { "Using Vault for secret injection" }
                });
            }

            var plainTextSecrets = _resources.Where(r => 
                r.Kind == "ConfigMap" &&
                (string?)r.Spec["data"]?["password"] != null ||
                (string?)r.Spec["data"]?["api-key"] != null);

            if (plainTextSecrets.Any())
            {
                _findings.Add(new Finding
                {
                    ControlId = $"{HipaaControls.ACCESS_CONTROL}:AC-2",
                    ResourcePath = string.Join(", ", plainTextSecrets.Select(c => c.Name)),
                    Description = "Potential secrets in ConfigMaps. Consider moving to Secrets.",
                    Passes = false,
                    Evidence = new() { "Check: " + string.Join(", ", plainTextSecrets.Select(c => c.Name)) }
                });
            }
        }

        public void GenerateScorecard()
        {
            var totalControls = _findings.DistinctBy(f => f.ControlId).Count();
            var passing = _findings.Where(f => f.Passes).DistinctBy(f => f.ControlId).Count();
            
            Console.WriteLine($"\n=== HIPAA Business Associate Readiness Scorecard ===");
            Console.WriteLine($"Total Controls Evaluated: {totalControls}");
            Console.WriteLine($"Passing: {passing} | Failing: {totalControls - passing}");
            Console.WriteLine($"Score: {(double)passing / totalControls * 100:F2}%\n");

            var byCategory = _findings.GroupBy(f => f.Category).ToDictionary(g => g.Key, g => g.ToList());
            
            foreach (var cat in byCategory.OrderBy(kvp => kvp.Key))
            {
                Console.WriteLine($"--- {cat.Key} ---");
                foreach (var finding in cat.Value)
                {
                    var status = finding.Passes ? "[PASS]" : "[FAIL]";
                    Console.WriteLine($"  {status} {finding.ControlId}: {finding.Description}");
                }
            }

            if (_findings.Any(f => !f.Passes))
            {
                Console.WriteLine("\n=== Priority Remediation Items ===");
                var failing = _findings.Where(f => !f.Passes).OrderByDescending(f => f.Category == HipaaControls.ACCESS_CONTROL);
                foreach (var item in failing.Take(5))
                {
                    Console.WriteLine($"  • {item.ControlId}: {item.Description}");
                }
            }
        }

        public static void Main(string[] args)
        {
            // Demo: Sample HIPAA manifest scan
            var scanner = new HipaaScanner();

            // Sample manifests (real-world would read from files/cluster)
            string[] samples = new[]
            {
                @"---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: app-sa
spec:
  automountServiceAccountToken: true",
                
                @"---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    audit-log-enabled: ""true""
spec:
  rules:
  - host: api.example.com
    http:
      paths:
      - path: /
        backend:
          service:
            name: api
            port:
              number: 80",

                @"---
apiVersion: v1
kind: ConfigMap
metadata:
  name: app-config
data:
  database-url: ""postgres://user:pass@db/""
  api-key: ""sk-1234567890""",

                @"---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: admin-binding
subjects:
- kind: User
  name: john-doe
roleRef:
  kind: ClusterRole
  name: cluster-admin
  apiGroup: rbac.authorization.k8s.io"
            };

            foreach (var sample in samples)
                scanner.AddManifest(sample);

            // Run the scan
            scanner.Scan();

            // Generate and display scorecard
            scanner.GenerateScorecard();

            // Output detailed JSON for CI/CD pipelines
            var output = JsonSerializer.Serialize(scanner._findings, new JsonSerializerOptions 
            { 
                WriteIndented = true,
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase
            });
            
            Console.WriteLine($"\n=== Raw JSON Output ===");
            Console.WriteLine(output);

            // Exit with appropriate code for CI/CD
            int exitCode = scanner._findings.Any(f => !f.Passes) ? 1 : 0;
            Environment.Exit(exitCode);
        }
    }
}

File: polyglot/java/manifest_parser.java
package polyglot.java;

import org.yaml.snakeyaml.Yaml;
import org.yaml.snakeyaml.constructor.Constructor;
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.time.Instant;

/**
 * HIPAA Security Rule Gap Scanner for Infrastructure Manifests.
 * Produces a Business Associate readiness scorecard with weighted scoring.
 */
public class ManifestParser {

    private static final double WEIGHT_ACCESS = 25.0;
    private static final double WEIGHT_AUDIT = 15.0;
    private static final double WEIGHT_INTEGRITY = 15.0;
    private static final double WEIGHT_PHYSICAL = 15.0;
    private static final double WEIGHT_TRANSMISSION = 30.0;

    public static class Manifest {
        public String name;
        public Map<String, Object> metadata;
        public List<Map<String, Object>> resources;
        public Map<String, Object> spec;
        
        // Extracted fields for HIPAA evaluation
        private Set<String> users = new HashSet<>();
        private boolean hasEncryptionInTransit = false;
        private boolean hasNetworkPolicies = false;
        private boolean hasAuditLogging = false;
        private boolean hasMfa = false;
        private boolean hasAutoLogoff = false;
        private boolean hasRoleBasedAccess = false;
    }

    public static class Score {
        public int accessControl;
        public int auditControls;
        public int integrityControls;
        public int physicalProcedures;
        public int transmissionSecurity;
        
        public double totalScore; // 0-100
        public String overallStatus;

        public static final int MAX_SCORE = 100;
    }

    /**
     * Parse a manifest file and extract HIPAA-relevant fields.
     */
    public Manifest parseManifest(String path) throws IOException {
        Path p = Paths.get(path);
        String content = Files.readString(p).trim();

        // Detect format
        if (content.contains("apiVersion: v1") || content.contains("kind:")) {
            return parseKubernetes(content);
        } else if (content.contains("\"Type\"") || content.contains("AWSTemplateFormatVersion")) {
            return parseCloudFormation(content);
        } else {
            // Generic YAML fallback
            Yaml yaml = new Yaml();
            Manifest m = yaml.load(content);
            if (m instanceof Manifest) {
                return (Manifest) m;
            }
            throw new IOException("Unrecognized manifest format");
        }
    }

    private Manifest parseKubernetes(String content) throws IOException {
        Yaml yaml = new Yaml();
        
        // Handle multi-document YAML (common in K8s manifests)
        List<Object> docs = yaml.loadAll(content);
        
        Manifest m = new Manifest();
        for (Object doc : docs) {
            if (!(doc instanceof Map)) continue;
            
            Map<String, Object> map = (Map<String, Object>) doc;
            
            // Extract resource name
            if (map.containsKey("metadata")) {
                @SuppressWarnings("unchecked")
                Map<String, Object> meta = (Map<String, Object>) map.get("metadata");
                m.name = extractName(meta);
                
                // Extract users from service accounts or secrets
                extractUsers(map);
            } else if (map.containsKey("spec")) {
                @SuppressWarnings("unchecked")
                Map<String, Object> spec = (Map<String, Object>) map.get("spec");
                m.spec = spec;
                
                // Check for encryption in transit (network policies)
                checkNetworkPolicies(spec);
            }
        }

        return m;
    }

    private Manifest parseCloudFormation(String content) throws IOException {
        Yaml yaml = new Yaml();
        
        @SuppressWarnings("unchecked")
        Map<String, Object> root = (Map<String, Object>) yaml.load(content);
        
        if (!root.containsKey("Resources")) {
            throw new IOException("Invalid CloudFormation manifest");
        }

        Manifest m = new Manifest();
        m.name = extractName((Map<String, Object>) root.get("Metadata"));

        @SuppressWarnings("unchecked")
        List<Map<String, Object>> resources = (List<Map<String, Object>>) 
                root.getOrDefault("Resources", Collections.emptyList());

        for (Object r : resources) {
            if (!(r instanceof Map)) continue;
            
            @SuppressWarnings("unchecked")
            Map<String, Object> res = (Map<String, Object>) r;
            
            // Extract resource type and properties
            String type = (String) res.getOrDefault("Type", "");
            m.resources.add(res);

            if (type.contains("NetworkSecurityGroup")) {
                checkNetworkPolicies(res);
            } else if (type.contains("SecretsManager") || type.contains("KMS")) {
                m.hasEncryptionInTransit = true;
            } else if (type.contains("IAMRole") || type.contains("User")) {
                extractUsers(res);
            }
        }

        return m;
    }

    private String extractName(Map<String, Object> meta) {
        if (meta.containsKey("name")) {
            return (String) meta.get("name");
        } else if (meta.containsKey("labels")) {
            @SuppressWarnings("unchecked")
            Map<String, Object> labels = (Map<String, Object>) meta.get("labels");
            String name = (String) labels.getOrDefault("app.kubernetes.io/name", "");
            return !name.isEmpty() ? name : "unnamed";
        }
        return "unknown";
    }

    private void extractUsers(Map<String, Object> map) {
        // Check for service accounts
        if (map.containsKey("spec") && 
            ((Map<String, Object>) map.get("spec")).containsKey("serviceAccountName")) {
            String sa = (String) ((Map<String, Object>) map.get("spec")).get("serviceAccountName");
            users.add(sa);
        }

        // Check for secrets containing credentials
        if (map.containsKey("data") && 
            ((Map<String, Object>) map.get("data")).containsKey("password")) {
            @SuppressWarnings("unchecked")
            Map<String, Object> data = (Map<String, Object>) map.get("data");
            users.add(data.keySet().iterator().next()); // The secret name
        }

        // Check for IAM roles/users in CloudFormation
        if (map.containsKey("Properties")) {
            @SuppressWarnings("unchecked")
            Map<String, Object> props = (Map<String, Object>) map.get("Properties");
            
            if (props.containsKey("UserName")) {
                users.add(props.get("UserName").toString());
            } else if (props.containsKey("RoleName")) {
                users.add(props.get("RoleName").toString());
            }
        }
    }

    private void checkNetworkPolicies(Map<String, Object> spec) {
        // Check for NetworkPolicy resources in K8s
        if (spec.containsKey("kind") && 
            "NetworkPolicy".equals(spec.get("kind"))) {
            hasNetworkPolicies = true;
        }

        // Check for VPC/Security Group properties
        if (spec.containsKey("VpcId") || spec.containsKey("SecurityGroupIds")) {
            hasNetworkPolicies = true;
        }
    }

    /**
     * Evaluate HIPAA Security Rule compliance.
     */
    public Score evaluate(Manifest manifest) {
        Score s = new Score();

        // 1. Access Control (25%)
        s.accessControl = evaluateAccessControl(manifest);

        // 2. Audit Controls (15%)
        s.auditControls = evaluateAuditControls(manifest);

        // 3. Integrity Controls (15%)
        s.integrityControls = evaluateIntegrityControls(manifest);

        // 4. Physical Procedures (15%)
        s.physicalProcedures = evaluatePhysicalProcedures(manifest);

        // 5. Transmission Security (30%)
        s.transmissionSecurity = evaluateTransmissionSecurity(manifest);

        // Calculate weighted total score
        double weightedSum = 
            s.accessControl * WEIGHT_ACCESS +
            s.auditControls * WEIGHT_AUDIT +
            s.integrityControls * WEIGHT_INTEGRITY +
            s.physicalProcedures * WEIGHT_PHYSICAL +
            s.transmissionSecurity * WEIGHT_TRANSMISSION;

        s.totalScore = Math.round(weightedSum / 100.0);

        // Determine overall status
        if (s.totalScore >= 90) {
            s.overallStatus = "READY";
        } else if (s.totalScore >= 75) {
            s.overallStatus = "NEEDS_REVIEW";
        } else if (s.totalScore >= 60) {
            s.overallStatus = "AT_RISK";
        } else {
            s.overallStatus = "CRITICAL";
        }

        return s;
    }

    private int evaluateAccessControl(Manifest manifest) {
        int score = 25; // Maximum for this category
        
        if (manifest.hasMfa) {
            score += 10;
        }
        
        if (manifest.hasAutoLogoff) {
            score += 8;
        }

        if (manifest.hasRoleBasedAccess) {
            score += 7;
        }

        // Check for unique user identification
        if (!manifest.users.isEmpty()) {
            score += 5;
        }

        return Math.min(score, 100);
    }

    private int evaluateAuditControls(Manifest manifest) {
        int score = 15;
        
        if (manifest.hasAuditLogging) {
            score += 12;
        }

        // Check for audit-related resources
        for (Object r : manifest.resources) {
            @SuppressWarnings("unchecked")
            Map<String, Object> res = (Map<String, Object>) r;
            
            if (res.containsKey("Type")) {
                String type = ((String) res.get("Type")).toLowerCase();
                if (type.contains("cloudwatchlogs") || 
                    type.contains("audittrail") ||
                    type.contains("s3" + "eventbridge")) {
                    score += 3;
                }
            }
        }

        return Math.min(score, 100);
    }

    private int evaluateIntegrityControls(Manifest manifest) {
        int score = 15;
        
        // Check for checksums or versioning
        if (manifest.spec != null && 
            ((Map<String, Object>) manifest.spec).containsKey("checksum")) {
            score += 8;
        }

        // Check for immutable infrastructure patterns
        if (manifest.resources != null) {
            for (Object r : manifest.resources) {
                @SuppressWarnings("unchecked")
                Map<String, Object> res = (Map<String, Object>) r;
                
                if (res.containsKey("Type")) {
                    String type = ((String) res.get("Type")).toLowerCase();
                    if (type.contains("efs") || 
                        type.contains("s3" + "versioning") ||
                        type.contains("snapshot")) {
                        score += 4;
                    }
                }
            }
        }

        return Math.min(score, 100);
    }

    private int evaluatePhysicalProcedures(Manifest manifest) {
        int score = 15;
        
        // Check for VPC isolation (physical/network boundary)
        if (manifest.hasNetworkPolicies) {
            score += 8;
        }

        // Check for dedicated environments
        if (manifest.spec != null && 
            ((Map<String, Object>) manifest.spec).containsKey("environment")) {
            String env = (String) ((Map<String, Object>) manifest.spec).get("environment");
            if ("production".equals(env) || "prod".equals(env)) {
                score += 7;
            }
        }

        return Math.min(score, 100);
    }

    private int evaluateTransmissionSecurity(Manifest manifest) {
        int score = 30; // Maximum for this category
        
        if (manifest.hasEncryptionInTransit) {
            score += 15;
        }

        // Check for TLS/SSL configuration
        if (manifest.spec != null && 
            ((Map<String, Object>) manifest.spec).containsKey("tls")) {
            @SuppressWarnings("unchecked")
            Map<String, Object> tls = (Map<String, Object>) ((Map<String, Object>) manifest.spec).get("tls");
            
            if (tls.containsKey("version") || tls.containsKey("minVersion")) {
                score += 10;
            }
        }

        // Check for secure ingress patterns
        if (manifest.resources != null) {
            for (Object r : manifest.resources) {
                @SuppressWarnings("unchecked")
                Map<String, Object> res = (Map<String, Object>) r;
                
                if (res.containsKey("Type")) {
                    String type = ((String) res.get("Type")).toLowerCase();
                    if (type.contains("ingress" + "ssl") || 
                        type.contains("alb" + "listener" + "tls")) {
                        score += 5;
                    }
                }
            }
        }

        return Math.min(score, 100);
    }

    /**
     * Generate a human-readable report.
     */
    public String generateReport(Manifest manifest, Score score) {
        StringBuilder sb = new StringBuilder();
        
        sb.append("HIPAA Security Rule Assessment Report\n");
        sb.append("======================================\n");
        sb.append(String.format("Timestamp: %s\n", Instant.now().toString()));
        sb.append(String.format("Manifest: %s\n", manifest.name));
        sb.append("\n");

        // Overall summary
        sb.append(String.format("OVERALL STATUS: %s (%.1f%%)\n", 
                score.overallStatus, score.totalScore));
        sb.append("\n");

        // Detailed breakdown
        sb.append("CATEGORY BREAKDOWN\n");
        sb.append("------------------\n");
        
        String[] categories = {
            "Access Control (25%)",
            "Audit Controls (15%)",
            "Integrity Controls (15%)",
            "Physical Procedures (15%)",
            "Transmission Security (30%)"
        };

        int[] scores = {
            score.accessControl,
            score.auditControls,
            score.integrityControls,
            score.physicalProcedures,
            score.transmissionSecurity
        };

        for (int i = 0; i < categories.length; i++) {
            String status = "PASS";
            if (scores[i] < 75) status = "WARN";
            if (scores[i] < 50) status = "FAIL";
            
            sb.append(String.format("  %-32s: %d/100 [%s]\n", 
                    categories[i], scores[i], status));
        }

        // Findings summary
        sb.append("\nKEY FINDINGS\n");
        sb.append("------------\n");

        List<String> findings = new ArrayList<>();

        if (manifest.hasEncryptionInTransit) {
            findings.add("+ Encryption in transit detected");
        } else {
            findings.add("- No explicit encryption-in-transit configuration found");
        }

        if (manifest.hasNetworkPolicies) {
            findings.add("+ Network policies configured");
        } else {
            findings.add("- Limited network segmentation observed");
        }

        if (!manifest.users.isEmpty()) {
            sb.append(String.format("  Users identified: %d\n", manifest.users.size()));
            for (String u : manifest.users) {
                findings.add(String.format("    * %s", u));
            }
        } else {
            findings.add("- No user/service account inventory found");
        }

        if (manifest.hasAuditLogging) {
            findings.add("+ Audit logging enabled");
        } else {
            findings.add("- Audit logging not explicitly configured");
        }

        // Recommendations
        sb.append("\nRECOMMENDATIONS\n");
        sb.append("---------------\n");

        if (score.totalScore < 90) {
            findings.add(String.format("1. Address %d critical/medium gaps before production deployment", 
                    100 - score.totalScore));
            
            if (!manifest.hasEncryptionInTransit) {
                findings.add("   * Configure TLS for all ingress traffic");
            }
            if (manifest.users.isEmpty()) {
                findings.add("   * Document all service accounts and users");
            }
        } else {
            findings.add(String.format("1. System is ready for HIPAA compliance review (%d%% score)", 
                    score.totalScore));
        }

        sb.append("\n");
        for (String f : findings) {
            sb.append(f).append("\n");
        }

        return sb.toString();
    }

    /**
     * Main entry point with demo functionality.
     */
    public static void main(String[] args) throws Exception {
        // Demo: Parse a sample Kubernetes manifest
        String sampleManifest